### PR TITLE
Fix perfornance issue of GroupNorm on CUDA when feature map is small.

### PR DESCRIPTION
**Pull request by xiaomengy**
Summary: Fix perfornance issue of GroupNorm on CUDA when feature map is small.

Test Plan: buck test mode/dev-nosan //caffe2/test:nn -- "GroupNorm"

Differential Revision: D24242738

As mentioned in https://github.com/pytorch/pytorch/issues/46086, the current GroupNorm implementation performs bad on CUDA when the feature map is small even compared to the impl via BatchNorm before PyTorch 1.5.1. This PR fixed the performance issue when the feature map is small.

Benchmark script:

```
import torch
import torch.nn.functional as F

from timeit import Timer

norm = torch.nn.GroupNorm(8, 512).cuda()

num = 5000

sizes = [(1024, 512, 14, 14), (1024, 512, 7, 7), (1024, 512)]


def forward(x):
    _ = norm(x)
    torch.cuda.synchronize()


def backward(y, grad):
    y.backward(grad, retain_graph=True)
    torch.cuda.synchronize()


if __name__ == "__main__":
    # warm up
    x = torch.rand(*(sizes[0]), dtype=torch.float,
                   device="cuda", requires_grad=True)
    for _ in range(100):
        forward(x)

    for size in sizes:
        x = torch.rand(*size, dtype=torch.float,
                       device="cuda", requires_grad=True)
        t = Timer("forward(x)", "from __main__ import forward, x")
        print(f"size = {size}:")
        t1 = t.timeit(num) / num * 1e6
        print(f"avg_forward_time =  {t1}us")

        y = norm(x)
        grad = torch.randn_like(y)
        t = Timer("backward(y, grad)", "from __main__ import backward, y, grad")
        t2 = t.timeit(num) / num * 1e6
        print(f"avg_backward_time = {t2}us")
```

Benchmark result after this PR on a V100 devgpu:
```
size = (1024, 512, 14, 14):
avg_forward_time =  1635.6191572034732us
avg_backward_time = 4140.7730475999415us
size = (1024, 512, 7, 7):
avg_forward_time =  463.6513736099005us
avg_backward_time = 1641.7451039887965us
size = (1024, 512):
avg_forward_time =  66.59087920561433us
avg_backward_time = 128.6882139975205us
```

Benchmark result before this PR on a V100 devgpu:
```
size = (1024, 512, 14, 14):
avg_forward_time =  1636.729855206795us
avg_backward_time = 5488.682465581223us
size = (1024, 512, 7, 7):
avg_forward_time =  465.88476160541177us
avg_backward_time = 3129.9425506033003us
size = (1024, 512):
avg_forward_time =  96.90486900508404us
avg_backward_time = 2319.4099438143894us
```

Run this benchmark script on PyTorch 1.5.1 (the build options may be different, just for reference):
```
size = (1024, 512, 14, 14):
avg_forward_time =  2728.9786524139345us
avg_backward_time = 9711.360842408612us
size = (1024, 512, 7, 7):
avg_forward_time =  773.7861637957394us
avg_backward_time = 2496.8661199789494us
size = (1024, 512):
avg_forward_time =  173.00677900202572us
avg_backward_time = 188.9778567943722us

```

### CODE REVIEW

File: aten/src/ATen/native/cuda/group_norm_kernel.cu [hunk 1]
```diff
@@ -74,7 +78,32 @@ __global__ void ComputeFusedParamsCUDAKernel(
 }
 
 template <typename T>
-__global__ void GroupNormForwardSimpleCUDAKernel(
+__global__ void GroupNorm1dForwardCUDAKernel(
```
Comment (ngimel) [inline aten/src/ATen/native/cuda/group_norm_kernel.cu:81]: I wonder if these pw ops wouldn't be better with TensorIterator - it has unrolled loads,

File: aten/src/ATen/native/cuda/group_norm_kernel.cu [hunk 2]
```diff
@@ -149,6 +183,199 @@ __global__ void ComputeGradOutputCoeffientCUDAKernel(
...
+  __shared__ T_ACC b_shared[kReduceTileSize][kReduceTileSize + 1];
+  const int64_t c = blockIdx.x * blockDim.x + threadIdx.x;
+  T_ACC dg_sum1 = 0;
+  T_ACC dg_sum2 = 0;
+  T_ACC db_sum1 = 0;
+  T_ACC db_sum2 = 0;
+  if (c < C) {
+    const int64_t G = group;
+    const int64_t D = C / G;
+    for (int64_t n = threadIdx.y; n < N; n += blockDim.y * 2) {
```
Comment (ngimel) [inline aten/src/ATen/native/cuda/group_norm_kernel.cu:317]: can you please leave brief comments on what this kernel is doing? Like, you unroll by a factor of 2, half of shared memory is to keep the first part of unrolled loads, half to keep the second etc (would it make sense to give another dimension to shared mem?)

File: aten/src/ATen/native/cuda/group_norm_kernel.cu [hunk 3]
```diff
@@ -438,6 +682,55 @@ void GroupNormKernelImpl(
...
+  }
+  if (dgamma != nullptr || dbeta != nullptr) {
+    if (N <= 128) {
+      const int64_t B = (C + kCUDANumThreads - 1) / kCUDANumThreads;
+      GammaBeta1dBackwardCUDAKernel1<T><<<B, kCUDANumThreads, 0, cuda_stream>>>(
+          N, C, G, dY, X, mean, rstd, dgamma, dbeta);
+    } else {
+      const int64_t B = (C + kReduceTileSize - 1) / kReduceTileSize;
+      constexpr int kThreadX = kReduceTileSize;
+      constexpr int kThreadY = kReduceTileSize / 2;
```
Comment (ngimel) [inline aten/src/ATen/native/cuda/group_norm_kernel.cu:725]: please leave a comment here (or assert) to prevent someone from changing block dimensions, because otherwise shared memory sizes won't work.

File: torch/testing/_internal/common_nn.py [hunk 4]
```diff
@@ -1582,6 +1582,24 @@ def fractional_max_pool3d_test(test_case):
         check_eval=True,
         desc='1d_affine',
     ),
+    dict(
+        module_name='GroupNorm',
```
Comment (ngimel) [inline torch/testing/_internal/common_nn.py:1586]: do the added test cases cover all the kernels that you've added?

**Review by ngimel — COMMENTED** [review]
That's a very nice performance improvement. It would be good to have the comments in the kernels and describe how you choose which path to use. Please make sure that you are testing all the kernel variants that you add (I'm not sure added tests cover everything). Also, can you enable and test bfloat16 dispatch on cuda? Hopefully it should just work, and we are enabling most bfloat16 operations now. 
Thanks again for the fix!

File: aten/src/ATen/native/cuda/group_norm_kernel.cu (hunk 2 above)
Comment (BIT-silence) [reply aten/src/ATen/native/cuda/group_norm_kernel.cu:171]: Done.

**Comment by BIT-silence** [comment]
> That's a very nice performance improvement. It would be good to have the comments in the kernels and describe how you choose which path to use. Please make sure that you are testing all the kernel variants that you add (I'm not sure added tests cover everything). Also, can you enable and test bfloat16 dispatch on cuda? Hopefully it should just work, and we are enabling most bfloat16 operations now.
> Thanks again for the fix!

BFloat16 has been enabled.

**Review by ngimel — APPROVED** [review]
Looks good, thank you!
